manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,639 @@
+        //In that way, we know that the state change happened between the new and previous values.
+
+        //TODO: Understand what happens with negate(), does it change states to the object itself?
+        boolean newValueTest =      pred.value.asPredicate(timerService::getCurrentTimeMillis).test(newValue        .getValue().get());
+        boolean previousValueTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(previousValue   .getValue().get());
+        //If the predicate fails, then no changes need to happen.
+
+        // newValue is not 1, previousValue == 1
+        if(!(!newValueTest && previousValueTest)) {
+            return Optional.empty();
+        }
+
+        // Grab all datapoints (To be replaced by AssetDatapointValueQuery)
+        List<AssetDatapoint> valueDatapoints = AssetDatapointService.getDatapoints(ref);
+        ArrayList<AssetDatapoint> list = new ArrayList<>(valueDatapoints);
+
+        // If there are no historical data found, add some first
+        if(list.isEmpty()) return Optional.empty();
+
+        //What we do now is, we will try to figure out the latest datapoint where the predicate fails,
+        //before the newValue.
+        //This means that, the state change took place between the datapoint we just found and its next one.
+
+        //Find the first datapoint that passes the negated predicate
+
+        AssetDatapoint StateChangeAssetDatapoint = null;
+
+        try {
+            for (int i = 0; i < list.size()-1; i++) {
+                // Not using Object.equals, but Datapoint.equals
+
+                AssetDatapoint currentDp = list.get(i);
+                AssetDatapoint theVeryPreviousDp = list.get(i+1);
+
+    //            So, if the currentDp passes the predicate,
+                boolean currentDpTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(currentDp.getValue());
+    //            and if the very previous one (NEXT one in the array and PREVIOUS in the time dimension).
+    //            FAILS the predicate,
+                boolean previousDpTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(theVeryPreviousDp.getValue());
+    //            A state change happened where the state we are looking for was turned on.
+    //            We want the currentDp.
+
+
+                if(currentDpTest && !previousDpTest){
+                    StateChangeAssetDatapoint = currentDp;
+                    break;
+                }
+            }
+
+            if (StateChangeAssetDatapoint != null){
+                if (!pred.value.asPredicate(timerService::getCurrentTimeMillis).test(StateChangeAssetDatapoint.getValue())){
+                    throw new Exception("Found state change datapoint failed predicate");
+                }
+            }else{
+                throw new Exception("Couldn't find asset state change value");
```

Comment:
I know it's being catched by line 529, but throwing a more specific exception type helps readability / future cases.